### PR TITLE
Section API

### PR DESCRIPTION
**Pull request by TPGamesNL**
### Description
Adds a Section API to Skript, to be used internally and by addons. 

The conditionals (if, else if and else), loops and while loops have been converted to the new system, but no other sections have been added.

---
**Target Minecraft Versions:** any
**Requirements:** none
**Related Issues:** #443, #1402 and #1449, #2118, #3680, #4000, #4005 (all non-fixing)

### CODE REVIEW

File: src/main/java/ch/njol/skript/lang/util/CodeSection.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+			currentSections.remove(currentSections.size() - 1);
+		}
+	}
+
+	/**
+	 * Loads the code using {@link #loadCode(SectionNode)}, while also
+	 * adjusting {@code hasDelayBefore} from {@link ParserInstance}, expecting the loaded code to be called
+	 * 0 or more times during execution of this section.
+	 */
+	protected void loadOptionalCode(SectionNode sectionNode) {
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/lang/util/CodeSection.java:57]: I don't completely understand when I should use loadCode vs loadOptionalCode. Could you explain a little more? ;D

File: src/main/java/ch/njol/skript/sections/ConditionalSection.java [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+			lastIf = null;
+		}
+
+		hadDelayBefore = getParser().getHasDelayBefore();
+		loadCode(sectionNode);
+		hasDelayAfter = getParser().getHasDelayBefore();
+		if (hadDelayBefore.isTrue())
+			return true;
+
+		if (type == ConditionalType.ELSE) {
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/sections/ConditionalSection.java:91]: some comments here could be useful I think

File: src/main/java/ch/njol/skript/sections/LoopSection.java [hunk 3]
```diff
@@ -33,52 +40,52 @@
...
- * A trigger section which represents a loop.
- * 
- * @author Peter Güttinger
- */
-public class Loop extends TriggerSection {
+public class LoopSection extends CodeSection {
 
-	private final Expression<?> expr;
+	static {
+		Skript.registerSection(LoopSection.class, "loop %objects%");
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/sections/LoopSection.java:46]: Does https://github.com/SkriptLang/Skript/blob/master/src/main/java/ch/njol/skript/expressions/ExprTimes.java need to be edited too?

**Review by APickledWalrus — COMMENTED** [review]
Nothing tooooo bad. Another reviewer is definitely needed to double check ;D

File: src/main/java/ch/njol/skript/sections/LoopSection.java (hunk 3 above)
Comment (TPGamesNL) [reply src/main/java/ch/njol/skript/sections/LoopSection.java:46]: I don't think so

File: src/main/java/ch/njol/skript/sections/ConditionalSection.java (hunk 2 above)
Comment (TPGamesNL) [reply src/main/java/ch/njol/skript/sections/ConditionalSection.java:93]: I added some comments, but I also changed the code a bit

File: src/main/java/ch/njol/skript/lang/util/CodeSection.java (hunk 1 above)
Comment (TPGamesNL) [reply src/main/java/ch/njol/skript/lang/util/CodeSection.java:59]: I added more Javadocs

File: src/main/java/ch/njol/skript/sections/ConditionalSection.java [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+		ELSE, ELSE_IF, IF
+	}
+
+	private ConditionalType type;
+	private Condition condition;
+
+	private Kleenean hasDelayAfter;
+
+	@Override
+	public boolean init(Expression<?>[] exprs,
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/sections/ConditionalSection.java:55]: Can all of the parameters go on one line like in `LoopSection`, or should they be split up there too?

**Review by APickledWalrus — APPROVED** [review]
Just that one question, but it's certainly not going to stop me from approving this! Nice work!

File: src/main/java/ch/njol/skript/effects/EffContinue.java [hunk 5]
```diff
@@ -48,8 +47,19 @@
...
 	}
 
-	@SuppressWarnings("null")
-	private Loop loop;
+	@SuppressWarnings("NotNullFieldNotInitialized")
+	private LoopSection loop;
+
+	@Override
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		List<LoopSection> loops = getParser().getCurrentLoops();
```
Comment (Mwexim) [inline src/main/java/ch/njol/skript/effects/EffContinue.java:55]: While this may be stuff for another PR, I feel like adding support for the while section is very easy (unless there needs to be an explicit discussion, which I would understand).

File: src/main/java/ch/njol/skript/lang/parser/ParserInstance.java [hunk 6]
```diff
@@ -116,8 +115,34 @@ public SkriptEvent getCurrentSkriptEvent() {
...
+	@Nullable
+	public <T extends TriggerSection> T getCurrentSection(Class<T> sectionClass) {
+		for (TriggerSection triggerSection : currentSections) {
+			if (sectionClass.isInstance(triggerSection))
+				return (T) triggerSection;
+		}
+		return null;
+	}
+
+	public List<LoopSection> getCurrentLoops() {
```
Comment (Mwexim) [inline src/main/java/ch/njol/skript/lang/parser/ParserInstance.java:145]: Why would there need to be this method when we could just use `getCurrentSection(LoopSection.class)`?
I would deprecate the method, and invoke that method instead to not break addons immediately.

File: src/main/java/ch/njol/skript/lang/util/CodeSection.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+import ch.njol.skript.lang.TriggerSection;
+import ch.njol.skript.lang.parser.ParserInstance;
+import ch.njol.util.Kleenean;
+
+import java.util.List;
+
+/**
+ * A {@link Section} with code contents.
+ */
+public abstract class CodeSection extends Section {
```
Comment (Mwexim) [inline src/main/java/ch/njol/skript/lang/util/CodeSection.java:33]: I feel like this class does not serve many purposes? I don't seem to understand why this inheritance is needed, instead of just adding it to the Section class instead. Could you provide me with some details perhaps?

File: src/main/java/ch/njol/skript/sections/ConditionalSection.java [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+import ch.njol.skript.lang.util.CodeSection;
+import ch.njol.util.Kleenean;
+import org.bukkit.event.Event;
+import org.eclipse.jdt.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.List;
+
+@SuppressWarnings("NotNullFieldNotInitialized")
+public class ConditionalSection extends CodeSection {
```
Comment (Mwexim) [inline src/main/java/ch/njol/skript/sections/ConditionalSection.java:36]: This is just my opinion, but I feel like sections need to follow the name convention. At skript-parser, we decided to opt for `SecLoop` and so on and so forth. This would need a small discussion first, and is just a consistency issue.

*This applies to all section class as well.*

File: src/main/java/ch/njol/skript/sections/ConditionalSection.java [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+
+	static {
+		Skript.registerSection(ConditionalSection.class,
+			"else",
+			"else if <.+>",
+			"[(1¦if)] <.+>");
```
Comment (Mwexim) [inline src/main/java/ch/njol/skript/sections/ConditionalSection.java:42]: I feel like the conditional section needs its own parsing. This means that before parsing the registered section, you parse the conditional one first (it still extends the CodeSection class to preserve the new utility methods).

What you are creating here is essentially a `<.+>` pattern. I don't particularly like that, and I think this is the only reasonable solution. It is used so often that it could perhaps hurt performance?

File: src/main/java/ch/njol/skript/sections/LoopSection.java [hunk 10]
```diff
@@ -33,52 +40,57 @@
 import java.util.Map;
 import java.util.WeakHashMap;
 
-/**
- * A trigger section which represents a loop.
- * 
- * @author Peter Güttinger
- */
-public class Loop extends TriggerSection {
+public class LoopSection extends CodeSection {
```
Comment (Mwexim) [inline src/main/java/ch/njol/skript/sections/LoopSection.java:43]: *(Naming conventions)*

File: src/main/java/ch/njol/skript/sections/WhileSection.java [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+import ch.njol.skript.lang.SkriptParser.ParseResult;
+import ch.njol.skript.lang.TriggerItem;
+import ch.njol.skript.lang.util.CodeSection;
+import ch.njol.util.Kleenean;
+import org.bukkit.event.Event;
+import org.eclipse.jdt.annotation.Nullable;
+
+import java.util.List;
+
+public class WhileSection extends CodeSection {
```
Comment (Mwexim) [inline src/main/java/ch/njol/skript/sections/WhileSection.java:34]: *(Naming conventions)*

**Review by Mwexim — CHANGES_REQUESTED** [review]
Lovely changes, something Skript needed for a long time. One issue that will remain is that it will need to wait for #3838 to be merged. After that, the conditions can be implemented smoothly with this new API.

File: src/main/java/ch/njol/skript/lang/util/CodeSection.java (hunk 7 above)
Comment (TPGamesNL) [reply src/main/java/ch/njol/skript/lang/util/CodeSection.java:33]: Originally the section was meant to automatically load the code in init, but I changed that behaviour later, so I'll remove the class now

File: src/main/java/ch/njol/skript/effects/EffContinue.java (hunk 5 above)
Comment (TPGamesNL) [reply src/main/java/ch/njol/skript/effects/EffContinue.java:55]: I also feel like this would fit better in a different PR, you should open an issue for it

File: src/main/java/ch/njol/skript/lang/parser/ParserInstance.java (hunk 6 above)
Comment (TPGamesNL) [reply src/main/java/ch/njol/skript/lang/parser/ParserInstance.java:145]: I removed the method, mainly because it is only present in the first alpha, so not many addons will have seen it, and I don't think any addons would need the method in the first place.

File: src/main/java/ch/njol/skript/sections/ConditionalSection.java (hunk 9 above)
Comment (TPGamesNL) [reply src/main/java/ch/njol/skript/sections/ConditionalSection.java]: I don't think this is needed, because there is no disadvantage to creating an all-matching pattern. It wouldn't make a performance difference either, as it will currently try parsing every SectionNode's key as a condition, and so would your method. 
I also don't feel like adding a separate system specifically for conditional parsing, even if they're a big part of the language.

**Comment by TPGamesNL** [comment]
> Lovely changes, something Skript needed for a long time. One issue that will remain is that it will need to wait for #3838 to be merged. After that, the conditions can be implemented smoothly with this new API.

Thanks. I don't think this PR should wait for #3838.

File: src/main/java/ch/njol/skript/sections/ConditionalSection.java (hunk 9 above)
Comment (Mwexim) [reply src/main/java/ch/njol/skript/sections/ConditionalSection.java]: If it doesn’t hurt perfomance, I agree.

**Comment by Mwexim** [comment]
> Thanks. I don't think this PR should wait for #3838.

I mean it can be implemented later on as well. One of the two PRs will need to adapt to one another then.
